rfdetr/util/early_stopping.py

```
@@ -25,33 +25,48 @@ def __init__(self, model, patience=5, min_delta=0.001, use_ema=False, verbose=Tr
         
     def update(self, log_stats):
         """Update early stopping state based on epoch validation metrics"""
-        # Get the mAP value from the log stats
-        if self.use_ema and 'ema_test_coco_eval_bbox' in log_stats:
-            current_map = log_stats['ema_test_coco_eval_bbox'][0]
-        elif 'test_coco_eval_bbox' in log_stats:
-            current_map = log_stats['test_coco_eval_bbox'][0]
+        regular_map = None
+        ema_map = None
+        
+        if 'test_coco_eval_bbox' in log_stats:
+            regular_map = log_stats['test_coco_eval_bbox'][0]
+        
+        if 'ema_test_coco_eval_bbox' in log_stats:
+            ema_map = log_stats['ema_test_coco_eval_bbox'][0]
+        
+        current_map = None
+        if regular_map is not None and ema_map is not None:
+            if self.use_ema:
+                current_map = ema_map
+                metric_source = "EMA"
+            else:
+                current_map = max(regular_map, ema_map)
+                metric_source = "max(regular, EMA)"
+        elif ema_map is not None:
+            current_map = ema_map
+            metric_source = "EMA"
+        elif regular_map is not None:
+            current_map = regular_map
+            metric_source = "regular"
         else:
-            # No valid mAP metric found, skip early stopping check
+            if self.verbose:
+                print("Early stopping: No valid mAP metric found, skipping check")
```

this should probably just raise an exception